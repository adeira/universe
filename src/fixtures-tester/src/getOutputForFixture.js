// @flow

type OperationOutput = any | Promise<any>;

export default async function getOutputForFixture(
  input: string,
  operation: (input: string) => OperationOutput,
  file: string,
): Promise<any> {
  const shouldThrow = /\.error\.\w+$/.test(file);
  if (shouldThrow) {
    let result;
    try {
      result = await operation(input);
    } catch (error) {
      return `THROWN EXCEPTION:\n\n${error.toString()}`;
    }
    throw new Error(`Expected test file '${file}' to throw, but it passed:\n${result}`);
  } else {
    return operation(input);
  }
}
