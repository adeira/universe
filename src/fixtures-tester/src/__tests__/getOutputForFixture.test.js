// @flow

import getOutputForFixture from '../getOutputForFixture';

it('returns the operation result as expected', async () => {
  await expect(
    getOutputForFixture('OK', (input) => input.toLowerCase(), 'fixture.txt'),
  ).resolves.toBe('ok');

  await expect(
    getOutputForFixture('OK', (input) => Promise.resolve(input.toLowerCase()), 'fixture.txt'),
  ).resolves.toBe('ok');
});

it('fails on unexpected errors', async () => {
  await expect(
    getOutputForFixture(
      'OK',
      () => {
        throw new Error('ups');
      },
      'fixture.txt',
    ),
  ).rejects.toMatchInlineSnapshot(`[Error: ups]`);

  await expect(
    getOutputForFixture('OK', () => Promise.reject(new Error('ups')), 'fixture.txt'),
  ).rejects.toMatchInlineSnapshot(`[Error: ups]`);
});

it('throws when error is expected but none was thrown', async () => {
  await expect(
    getOutputForFixture('OK', (input) => input.toLowerCase(), 'fixture.error.txt'),
  ).rejects.toMatchSnapshot();

  await expect(
    getOutputForFixture('OK', (input) => Promise.resolve(input.toLowerCase()), 'fixture.error.txt'),
  ).rejects.toMatchSnapshot();
});

it('returns THROWN EXCEPTION when the operation is expected to fail and it fails', async () => {
  await expect(
    getOutputForFixture(
      'OK',
      () => {
        throw new Error('ups');
      },
      'fixture.error.txt',
    ),
  ).resolves.toMatchSnapshot();

  await expect(
    getOutputForFixture('OK', () => Promise.reject(new Error('ups')), 'fixture.error.txt'),
  ).resolves.toMatchSnapshot();
});
