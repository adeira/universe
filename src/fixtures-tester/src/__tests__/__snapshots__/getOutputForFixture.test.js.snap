// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns THROWN EXCEPTION when the operation is expected to fail and it fails 1`] = `
"THROWN EXCEPTION:

Error: ups"
`;

exports[`returns THROWN EXCEPTION when the operation is expected to fail and it fails 2`] = `
"THROWN EXCEPTION:

Error: ups"
`;

exports[`throws when error is expected but none was thrown 1`] = `
[Error: Expected test file 'fixture.error.txt' to throw, but it passed:
ok]
`;

exports[`throws when error is expected but none was thrown 2`] = `
[Error: Expected test file 'fixture.error.txt' to throw, but it passed:
ok]
`;
